client/client.go

```diff
@@ -156,10 +156,136 @@ type tsoRequest struct {
+	// This loop is for trying best to collect more requests, so we use `tbc.maxBatchSize` here.
+	for tbc.collectedRequestCount < tbc.maxBatchSize {
+		if len(tbc.tsoRequestCh) == 0 {
+			return true
+		}
+		select {
+		case tsoReq := <-tbc.tsoRequestCh:
+			// No more pending TSO requests, just return.
+			if tsoReq == nil {
+				return true
+			}
+			tbc.pushRequest(tsoReq)
+		case <-ctx.Done():
+			return false
+		default:
+			return true
+		}
+	}
+	return true
+}
+
+func (tbc *tsoBatchController) pushRequest(tsoReq *tsoRequest) {
+	tbc.collectedRequests[tbc.collectedRequestCount] = tsoReq
+	tbc.collectedRequestCount++
+}
+
+// shouldFetchMorePendingRequests returns whether we should fetch more pending TSO requests from the channel.
+func (tbc *tsoBatchController) shouldFetchMorePendingRequests() bool {
+	// TODO: maybe consider the actual load that returns through a TSO response from PD server.
+	return tbc.collectedRequestCount < tbc.maxBatchSize && tbc.maxBatchWaitInterval > 0
+}
+
+// fetchMorePendingRequests fetches more pending TSO requests from the channel.
+// It returns true if everything goes well, otherwise false which should stop the service.
+func (tbc *tsoBatchController) fetchMorePendingRequests(ctx context.Context) bool {
+	// Try to collect `tbc.bestBatchSize` requests, or wait `tbc.maxBatchWaitInterval`
+	// when `tbc.collectedRequestCount` is less than the `tbc.bestBatchSize`.
+	if tbc.collectedRequestCount < tbc.bestBatchSize {
+		after := time.NewTimer(tbc.maxBatchWaitInterval)
+		defer after.Stop()
+		for tbc.collectedRequestCount < tbc.bestBatchSize {
+			select {
+			case tsoReq := <-tbc.tsoRequestCh:
+				tbc.pushRequest(tsoReq)
+			case <-ctx.Done():
+				return false
+			case <-after.C:
+				return true
+			}
+		}
+	}
+
+	// Do an additional non-block try. Here we test the length with `tbc.maxBatchSize` instead
+	// of `tbc.bestBatchSize` because trying best to fetch more requests is necessary so that
+	// we can adjust the `tbc.bestBatchSize` dynamically later.
+	for tbc.collectedRequestCount < tbc.maxBatchSize {
+		if len(tbc.tsoRequestCh) == 0 {
+			return true
+		}
+		select {
+		case tsoReq := <-tbc.tsoRequestCh:
+			tbc.pushRequest(tsoReq)
+		case <-ctx.Done():
+			return false
+		default:
+			return true
+		}
+	}
+	return true
+}
+
+// adjustBestBatchSize stabilizes the latency with the AIAD algorithm.
+func (tbc *tsoBatchController) adjustBestBatchSize() {
+	tsoBestBatchSize.Observe(float64(tbc.bestBatchSize))
+	length := tbc.collectedRequestCount
+	if length < tbc.bestBatchSize && tbc.bestBatchSize > 1 {
+		// Waits too long to collect requests, reduce the target batch size.
+		tbc.bestBatchSize--
+	} else if length > tbc.bestBatchSize+4 /* Hard-coded number, in order to make `tbc.bestBatchSize` stable */ &&
+		tbc.bestBatchSize < tbc.maxBatchSize {
+		tbc.bestBatchSize++
+	}
+}
+
+func (tbc *tsoBatchController) observeBatchSendLatency() {
+	tsoBatchSendLatency.Observe(float64(time.Since(tbc.batchStartTime)))
```

Comment:
Why not directly use `tsoBatchSendLatency.Observe(float64(time.Since(tbc.batchStartTime)))`?